helix-view/src/document.rs

```diff
@@ -139,14 +143,59 @@ pub fn canonicalize_path(path: &Path) -> std::io::Result<PathBuf> {
 use helix_lsp::lsp;
 use url::Url;
 
+// Temporary
+#[inline(always)]
+fn checked_add(num: usize, opt: Option<usize>) -> Option<usize> {
+    if let Some(n) = opt {
+        n.checked_add(num)
+    } else {
+        None
+    }
+}
+
+// Temporary
+#[inline(always)]
+fn checked_next_power_of_two(opt: Option<usize>) -> Option<usize> {
+    opt.map(|n| n.next_power_of_two())
+}
+
+// Temporary
+#[inline(always)]
+fn checked_min(one: Option<usize>, other: Option<usize>) -> Option<usize> {
+    if let Some(a) = one {
+        if let Some(b) = other {
+            Some(::core::cmp::min(a, b))
+        } else {
+            Some(a)
+        }
+    } else {
+        other
+    }
+}
+
 impl Document {
-    pub fn new(text: Rope) -> Self {
+    pub fn new() -> Self {
+        let text = Rope::from("\n");
+
+        Self::from(text, None, None)
+    }
+
+    pub fn from(
+        text: Rope,
+        encoding: Option<&'static encoding_rs::Encoding>,
+        file_encoding: Option<&'static encoding_rs::Encoding>,
+    ) -> Self {
         let changes = ChangeSet::new(&text);
         let old_state = None;
+        let encoding =
+            encoding.unwrap_or(encoding_rs::Encoding::for_label("utf-8".as_bytes()).unwrap());
+        let file_encoding = file_encoding.unwrap_or(encoding.clone());
```

Comment:
We only need encoding, because the in memory representation will always be valid UTF-8. The encoding setting is simply for transcoding from/to file.

Reply:
I agree with archseer here.

kirawi: correct me if I'm wrong, but I *think* the intent here is to handle separately the concern of how the file was loaded (e.g. maybe you get nonsense, and want to reload the file with a different encoding) vs how you want it to be saved (maybe you want to change the encoding just so it gets saved differently).  It that along the lines of what you were thinking?

If so, I'm right there with you.  But I think that should probably be handled at a command level.  For example, "reload_with_encoding" which fully reloads the file, decoding it differently, and "change_encoding" which just changes the encoding setting of the document (so next time it will get saved with that encoding).  Or something like that.  There probably doesn't need to be two fields in the document struct itself.

Reply:
@cessen I think vim allows you to use a different internal encoding for the in-memory representation (I assume this is largely historical and might predate Unicode), hence two commands:

>   'encoding' 'enc'
>         String-encoding used internally and for RPC communication.
>         Always UTF-8.
> 
>  See 'fileencoding' to control file-content encoding.

Reply:
IIRC, yeah, I think you're right.  I believe vim, emacs, etc. allow different in-memory encodings, which would justify two settings.

(As an aside: you're mostly right that it's historical.  But it also turns out there are encodings that Unicode can't fully round-trip.  And you don't even have to get especially esoteric.  Many (most?) common variants of BIG5, for example, can't *quite* fully round-trip in the general case.  So an editor that wants to perfectly support all of the world's text data can't just rely on Unicode, and has to implement support for some encodings on a case-by-case basis.  If Helix ends up really caring about that in the future, I'd definitely be down to help support that kind of thing in Ropey.  But it's honestly a huge can of worms, and IMO probably not worth the marginal gains.)

Reply:
Yeah, I think it would be useful if you were editing binary or something like that, so I'm dropping the feature.

Reply:
I changed my mind, as it seems like a straightforward integration and is a bit more useful.